submission/persistence-mongo/src/main/kotlin/ac/uk/ebi/biostd/persistence/doc/service/StatsMongoDataService.kt

```diff
@@ -1,62 +1,76 @@
+            ?: throw StatsNotFoundException(accNo)
+
     override fun findByType(submissionStatType: SubmissionStatType, filter: PaginationFilter): List<SubmissionStat> =
         statsDataRepository
             .findAllByStatType(submissionStatType, PageRequest.of(filter.pageNumber, filter.limit))
             .content
-            .map { toSubmissionStat(it, submissionStatType) }
+            .map { toSubmissionStat(submissionStatType, it) }
 
     override fun findByAccNoAndType(accNo: String, submissionStatType: SubmissionStatType): SubmissionStat =
         statsDataRepository
             .findByAccNoAndStatType(accNo, submissionStatType)
-            ?.let { toSubmissionStat(it, submissionStatType) }
+            ?.let { toSubmissionStat(submissionStatType, it) }
             ?: throw StatNotFoundException(accNo, submissionStatType)
 
-    // TODO support saving stats for expired versions Pivotal ID # 176802128
+    override fun save(stat: SubmissionStat): SubmissionStat {
+        updateOrRegister(stat)
+        return stat
+    }
+
     override fun saveAll(stats: List<SubmissionStat>): List<SubmissionStat> =
         stats.filterValid()
             .mapValues { it.value.last() }
-            .map { updateSubmissionStat(it.key!!, it.value) }
+            .map { updateOrRegister(it.value) }
+
 
     override fun incrementAll(stats: List<SubmissionStat>): List<SubmissionStat> =
         stats.filterValid()
-            .mapValues { incrementSubmissionStat(it.key!!, it.value) }
+            .mapValues { increment(it.key, it.value) }
             .values
             .toList()
 
     private fun List<SubmissionStat>.filterValid() =
         groupBy { it.accNo.uppercase() }
-            .mapKeys { statsDataRepository.findByAccNo(it.key.uppercase()) }
-            .filterKeys { it != null }
+            .filter { submissionsRepository.existsByAccNo(it.key) }
 
-    private fun updateSubmissionStat(sub: DocSubmission, update: SubmissionStat): SubmissionStat {
-        statsDataRepository.updateStat(sub.accNo, sub.version, update)
+    private fun updateOrRegister(update: SubmissionStat): SubmissionStat {
+        statsDataRepository.updateOrRegisterStat(update)
         return update
     }
 
-    private fun incrementSubmissionStat(sub: DocSubmission, increments: List<SubmissionStat>): SubmissionStat {
+    private fun increment(accNo: String, increments: List<SubmissionStat>): SubmissionStat {
         val type = increments.first().type
-        val current = sub.stats.find { it.name == type.name }?.value ?: 0L
-        val increment = statsDataRepository.incrementStat(sub.accNo, sub.version, increments)
+        val current = statsDataRepository.findByAccNo(accNo)?.stats?.get(type.name) ?: 0L
+        val increment = statsDataRepository.incrementStat(accNo, increments)
 
-        return SingleSubmissionStat(sub.accNo, current + increment, type)
+        return SingleSubmissionStat(accNo, current + increment, type)
```

Comment:
rather that computer return value I would just update and query, in that case we return what is in database rather than asuming it.